src/utils/typedefs-wrapper.cpp

```diff
@@ -44,26 +46,286 @@ namespace detail {
 }
 
+StringW::operator const std::u16string_view() const {
+    return {inst->chars, static_cast<std::size_t>(inst->length)};
+}
+
 StringW::operator std::u16string_view() {
     return {inst->chars, static_cast<std::size_t>(inst->length)};
 }
+
+bool strcomp(Il2CppString const* lhs, const std::u16string_view rhs) noexcept {
+    if (!lhs || lhs->length != (int)rhs.size()) return false;
+
+    Il2CppChar const* first = lhs->chars; 
+    auto* second = rhs.data(); 
+    Il2CppChar const* firstEnd = first + lhs->length; 
+    auto* secondEnd = second + (int)rhs.size(); 
+
+    while (first != firstEnd && second != secondEnd)
+    {
+        if (*first != *second) return false;
+        first++; second++;
+    }
+
+    return first == firstEnd && second == secondEnd;
+}
+
+bool strcomp(Il2CppString const* lhs, const std::string_view rhs) noexcept {
+    if (!lhs || lhs->length != (int)rhs.size()) return false;
+
+    Il2CppChar const* first = lhs->chars; 
+    auto* second = rhs.data(); 
+    Il2CppChar const* firstEnd = first + lhs->length; 
+    auto* secondEnd = second + (int)rhs.size(); 
+
+    while (first != firstEnd && second != secondEnd)
+    {
+        if (*first != *second) return false;
+        first++; second++;
+    }
+
+    return first == firstEnd && second == secondEnd;
+}
+
+bool strless(Il2CppString const* lhs, const std::string_view rhs) noexcept {
+    if (!lhs) return true;
+
+    Il2CppChar const* first = lhs->chars; 
+    char const* second = rhs.data(); 
+    Il2CppChar const* firstEnd = lhs->chars + lhs->length; 
+    char const* secondEnd = rhs.data() + (int)rhs.size(); 
+    
+    while (first != firstEnd && second != secondEnd)
+    {
+        if (*first == *second)
+        {
+            first++; second++;
+            continue;
+        }
+        return *first < *second;
+    }
+    // if we got here, and second is not second end, we had a shorter first, so it should be true
+    // if second is the end, we are longer, so it should be false
+    return second != secondEnd;
+}
+
+bool strless(Il2CppString const* lhs, const std::u16string_view rhs) noexcept {
+    if (!lhs) return true;
+
+    Il2CppChar const* first = lhs->chars; 
+    char16_t const* second = rhs.data(); 
+    Il2CppChar const* firstEnd = first + lhs->length; 
+    char16_t const* secondEnd = second + (int)rhs.size(); 
+    
+    while (first != firstEnd && second != secondEnd)
+    {
+        if (*first == *second)
+        {
+            first++; second++;
+            continue;
+        }
+        return *first < *second;
+    }
+    // if we got here, and second is not second end, we had a shorter first, so it should be true
+    // if second is the end, we are longer, so it should be false
+    return second != secondEnd;
+}
+
+StringW strappend(Il2CppString const* lhs, const std::string_view rhs) noexcept {
+    return std::u16string(lhs->chars, lhs->chars + lhs->length).append(std::u16string(rhs.data(), rhs.data() + rhs.size()));
```

Comment:
Consider using the conversion functions to ensure validity instead of using pointer values that are NOT correct (utf8 chars, but creating a string of utf16 chars. 3 characters would parse into a utf16 string of 1 character, where the character would be half one and half the other)

Reply:
Also note really a huge fan of the double allocation here anyways, the append should be a known size, so we should be able to only need to allocate space for ONE new Il2CppString* and copy all the chars to there element by element. This also feels like something that would be more performant if it could take advantage of the fact that the lhs could be a `StringW` and not an `Il2CppString`, since if it's an `Il2CppString`, we don't have the power to allocate as easily (though we still could just fine).

Basically, I think this function, and where it is used, could stand to avoid these two excess allocations, as well as incorrect behavior with `rhs` conversion. _Ideally_, this function should make a new Il2CppString* that is of the correct size, and copy each character in, promoting the `char` characters to `char16_t`. Hopefully that makes sense.